fix: Client errors during networked scene loads

NoelStephensUnity opened the pull request:
[MTT-1215](https://jira.unity3d.com/browse/MTT-1215)
[Github Issue 1102](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/issues/1102)
This fixes the bug as described in issue-1102 where under specific scenarios a server could respond back to the client before the client was finished loading a scene which would result in the client having its singular SceneEventData overwritten.  After looking at the possible fixes, it was determined that the best fix for this would be to have a SceneEventData pool to handle this condition.
This includes the ability to adjust the pool size in the event a user needs a dedicated server to handle many clients loading and/or unloading many additive scenes in a short period of time.

Added tests:
SceneEventDataPoolSceneLoadingTest
SceneEventDataPoolTests

NoelStephensUnity committed:
fix

This fixes Github issue 1102.
The issue was that the server was running faster than the disk I/O for scene loading and so the SceneEventData single instance was being written over before the client was done loading.
To avoid this issue and any future potential issues I created a SceneEventData pool and use a unique SceneEventData instance for each scene event.  Currently set the default pool size to 10 but made it so it can be adjusted if needed through the public facing method NetworkSceneManager.SetSceneEventDataPoolSize.

NoelStephensUnity committed:
refactor and style

Added an upper threshold that a user can set the SceneEventData pool size.
Adjusted some of the naming.
Exposed the default and upper threshold constants.

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

NoelStephensUnity committed:
fix

Making sure I clear out the list after disposing the pool.

NoelStephensUnity committed:
refactor and style

Applying some stylistic changes and going ahead and merging the squelching of validation warnings as it was duplicated.

NoelStephensUnity committed:
fix

No idea... weirdest thing ever.
Fixing "new ()"

NoelStephensUnity committed:
style

whitespace checks

NoelStephensUnity committed:
test

Added integration test for SceneEvenData pool

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

NoelStephensUnity committed:
test

Adding additional tests to check the bounds when setting the data pool size.

NoelStephensUnity committed:
style

fixing whitespace issues.

NoelStephensUnity committed:
refactor

Fixed some bounds checking (both test and in NetworkSceneManager) issues.

NoelStephensUnity committed:
style

updated comments
removed some LFs

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

NoelStephensUnity committed:
refactor

Fixing differences from previous merge.

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

NoelStephensUnity committed:
fix

fixes for the develop merge.

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -185,13 +180,34 @@ public class NetworkSceneManager : IDisposable
         /// not destroy temporary scene are moved into the active scene
         /// </summary>
-        internal static bool IsSpawnedObjectsPendingInDontDestroyOnLoad = false;
+        internal static bool IsSpawnedObjectsPendingInDontDestroyOnLoad;
```
Even if default assignment is not needed, here, wouldn't having a default value help people understand the code and remove a potential confusion on the default value?

NoelStephensUnity replied:
The default value of a bool is false and it is non-standard to set the property of a bool to something it already defaults to.

mattwalsh-unity replied:
Yeah, I was the same way @jeffreyrainy but eventually I've gotten habituated to C#'s defaults

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -472,12 +552,12 @@ private void SendSceneEventData(ulong[] targetClientIds)
 
             m_NetworkManager.NetworkMetrics.TrackSceneEventSent(
-                targetClientIds, (uint)SceneEventData.SceneEventType, ScenesInBuild[(int)SceneEventData.SceneIndex], size);
+                targetClientIds, (uint)SceneEventDataPool[sceneEventDataIndex].SceneEventType, ScenesInBuild[(int)SceneEventDataPool[sceneEventDataIndex].SceneIndex], size);
```
This was already present in the code before the change, but I'm curious. How come the typecast is not an indication of a potential improvement? Shouldn't `TrackSceneEventSent` already take a parameter of the type of `SceneEventData.SceneEventType` ? It seem it would be safer if both were the same type.

NoelStephensUnity replied:
Good question, that perhaps the tools team would be better suited to answer

kvassall-unity replied:
`INetworkMetrics` is type agnostic. Probably because we don't want to leak types out externally. (I say probably because I didn't write INetworkMetrics)

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

jeffreyrainy commented:
There's some areas I'm unfamiliar with. That makes me unwilling to approve, this close to the release, but I have no issue with the code. I also ran it locally and it seems to behave ok, with and without snapshot enabled for spawn.

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

NoelStephensUnity committed:
fix

Re-adding the UnitTestBaseScene back into the Build Settings Scenes in Build list.
This is required by NetworkSeneManager tests.

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -224,16 +240,75 @@ public class NetworkSceneManager : IDisposable
         /// </summary>
+        /// <param name="sceneEventDataPoolSize"></param>
         /// <returns></returns>
+        public bool SetSceneEventDataPoolSize(int sceneEventDataPoolSize)
```
Connected to our previous conversation, this seems like a fair bit of complexity and configurability for something that I don't think users will get into.

What I see here is, we have 3 settings which 
- if exceeded generate a warning
- but, the situation is worse than a warning - when the warning happens it actually has failed to do the operation.  That sounds more like an error to me.  Or somehow is the pool failing to grow still ok?

Again I think these settings add complexity and barring coming to an improved understanding would like to remove them.

Put another way, if your game is set up such that you blow through our expected maximums, just having a "you blew through our maximums so set them higher else we'll warn you and put you in an unexpected state" is not adding value.

If you want to convince me that we must give them a way to increase DefaultSceneEventDataPoolSize because whatever default we pick would if hard-coded cause harm to users because they'll have to waste cycles in their pool growing up to their larger amounts, we can have that conversation too.

NoelStephensUnity replied:
This was resolved by just dynamically generating SceneEventData objects per each unique scene event.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -282,11 +357,18 @@ public void SetClientSynchronizationMode(LoadSceneMode mode)
+            SceneEventDataPool = new List<SceneEventData>();
+            for (int i = 0; i < m_MaxSceneEventDataPoolSize; i++)
+            {
+                SceneEventDataPool.Add(new SceneEventData(networkManager));
```
I guess I find this odd that we would always warm up the pool with m_MaxSceneEventDataPoolSize allocations of SceneEventData.  Maybe they're not using scene changing at all?  Is the pool growth really that expensive?

NoelStephensUnity replied:
If a client connects you are using SceneEvents and as such you will be using the SceneEventDataPool.  I am not aware of any network based Unity created game that never loads any scenes during the network session.  Do we think this is possible?
I am willing to make all constants and pool size related methods internal only, but I do want to say that removing that option removes a user's ability to adjust that knob until we push an update.

mattwalsh-unity reviewed (changes requested):
I need to get more comfortable about all the parameter complexity before I can approve.  Also this is really hard to carefully scrub with the variable renames

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

NoelStephensUnity committed:
refactor

Switched to dynamically generated SceneEventData objects per every Scene Event.
Removed the SceneEventDataPool as well as the associated methods and properties.
Adjusted NetworkSceneManagerTests to reflect this change.

NoelStephensUnity committed:
style

Updated existing and missing comments.

NoelStephensUnity committed:
fix

adding missing scenes.

NoelStephensUnity committed:
fix

last scene didn't make last update.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -185,19 +180,14 @@ public class NetworkSceneManager : IDisposable
+        /// Used for all scene event processing
         /// </summary>
-        internal SceneEventData SceneEventData;
+        internal Dictionary<Guid, SceneEventData> SceneEventDataQueue;
```
"Queue" suffix implies `Queue<T>` by the name — what about `SceneEventDataStore` or something like that?

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -224,16 +214,46 @@ public class NetworkSceneManager : IDisposable
-            ClientSynchEventData.Dispose();
-            ClientSynchEventData = null;
+            foreach (var keypair in SceneEventDataQueue)
+            {
```
a (warning?) log here would be useful in case of leaving some events unprocessed. it could help us when we debug things.

NoelStephensUnity replied:
Yeah, I thought about that but there is a chance that someone just shutdowns the application in the middle of a scene event.
So, I guess I could at a minimum log SceneEventData object entries being disposed.

0xFA11 replied:
it could be lower than a warning log. we had something like `NetworkLogLevel.Develop` or `NetworkLogLevel.Debug`, that'd become handy here. minor enough, I'll leave it up to you.

NoelStephensUnity replied:
Actually, I like it at the normal level because in all reality the only time that should log anything is if you are exiting in the middle of a scene event.  Any other scenario it should not log anything as there should not be any events in the store.  As well, it isn't a warning...just informational that things are being disposed and removed.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/SceneEventData.cs:
```diff
@@ -766,6 +768,7 @@ public void Dispose()
         internal SceneEventData(NetworkManager networkManager)
         {
             m_NetworkManager = networkManager;
+            SceneEventId = Guid.NewGuid();
```
nit: instead of carrying `Guid` type around, it could be fancier to do a `.ToString()` on the `Guid` instance and then `XXHash` that `string` to convert it to an `int` hash value.

NoelStephensUnity replied:
Hmmm... I like your brain.

0xFA11 reviewed (approved):
some minors/nits but otherwise, looks good to me!

NoelStephensUnity committed:
refactor

Changed SceneEventDataQueue to SceneEventDataStore.
Changed SceneEvent.SceneEventId from guid to uint.

NoelStephensUnity committed:
update

Added logging of any remaining SceneEventData objects in the store.
Based on console output after this, added some additional EndSceneEvent calls in places that were missed due to early exits or just missed.

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

NoelStephensUnity committed:
Merge branch 'develop' into fix/github-issue-1102

mattwalsh-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/SceneEventData.cs:
```diff
@@ -90,7 +90,9 @@ public enum SceneEventTypes : byte
         internal LoadSceneMode LoadSceneMode;
-        internal Guid SceneEventGuid;
+        internal Guid SceneEventProgressId;
+        internal uint SceneEventId;
```
Nit, I think actually in this case it *is* a GUID, no?
